File: 백준/Silver/9012. 괄호/괄호.py
import sys
input=sys.stdin.readline


n=int(input())
for i in range(n):
    ps = input()
    cnt = 0
    for j in range(len(ps)):
        if ps[j] == '(':
            cnt+=1
        elif ps[j] == ')':
            cnt-=1
        if cnt<0:
            print("NO")
            break
    if cnt == 0:
        print("YES")
    elif cnt>0:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.js
const fs = require('fs');

fs.readFile('/dev/stdin', 'utf8', (err, data) => {
    if (err) {
        console.error(err);
        return;
    }
    
    let inputLines = data.trim().split('\n');
    let n = parseInt(inputLines[0]);
    
    for (let i = 1; i <= n; i++) {
        let ps = inputLines[i];
        let cnt = 0;
        let isValid = true;
        
        for (let j = 0; j < ps.length; j++) {
            if (ps[j] === '(') {
                cnt++;
            } else if (ps[j] === ')') {
                cnt--;
            }
            
            if (cnt < 0) {
                console.log("NO");
                isValid = false;
                break;
            }
        }
        
        if (isValid) {
            if (cnt === 0) {
                console.log("YES");
            } else {
                console.log("NO");
            }
        }
    }
});